src/amplify-migration/amplifyMigrationScript.ts

```diff
@@ -0,0 +1,289 @@
+        permalinkIndex,
+        fileContent.indexOf("\n", permalinkIndex)
+      );
+      let permalinkLineTrimmed = permalinkLine.trim();
+
+      let hasTrailingQuotation = false;
+      if (permalinkLineTrimmed.endsWith(`"`)) {
+        hasTrailingQuotation = true;
+        permalinkLineTrimmed = permalinkLineTrimmed.slice(0, -1);
+      }
+      const permalinkWithSlash = permalinkLineTrimmed.endsWith("/")
+        ? permalinkLineTrimmed
+        : `${permalinkLineTrimmed}/`;
+      let newPermalink = permalinkWithSlash.toLocaleLowerCase();
+      if (hasTrailingQuotation) {
+        newPermalink = `${newPermalink}"`;
+      }
+      const newFileContent = fileContent.replace(
+        permalinkLine.trim(),
+        newPermalink
+      );
+
+      await fs.promises.writeFile(filePath, newFileContent, "utf-8");
+      //remove trailing slash
+      newPermalink = getRawPermalink(newPermalink);
+      const originalPermalink = getRawPermalink(permalinkLine);
+      if (newPermalink !== originalPermalink) {
+        changedPermalinks[originalPermalink] = newPermalink;
+      }
+      await simpleGit(repoPath).add(filePath);
+    }
+  }
+  console.log(changedPermalinks);
+
+  for (const file of Object.values(mdFiles as {})) {
+    const filePath = path.join(repoPath, file);
+    await changePermalinksInFiles(filePath, repoPath, changedPermalinks);
+  }
+
+  const commitMessage =
+    "chore(Amplify-Migration): Update permalinks in .md files";
+  await simpleGit(repoPath).commit(commitMessage);
+}
+
+/**
+ * Requirements:
+ * convert `permalink: /some/path/` to `some/path`
+ * convert `/some/path` to `some/path`
+ * convert `some/path/` to `some/path`
+ * @param permalink original permalink
+ * @returns raw permalinks without leading/trailing slash
+ */
+function getRawPermalink(permalink: string) {
+  let trimmedPermalink = permalink.trim();
+  if (trimmedPermalink.startsWith(`permalink: `)) {
+    trimmedPermalink = permalink.trim().slice(11);
+  }
+  if (trimmedPermalink.startsWith(`/`)) {
+    trimmedPermalink = trimmedPermalink.slice(1);
+  }
+  if (trimmedPermalink.endsWith("/")) {
+    trimmedPermalink = trimmedPermalink.slice(0, -1);
+  }
+
+  return trimmedPermalink;
+}
+
+async function changePermalinksInFiles(
+  filePath: string,
+  repoPath: string,
+  changedPermalinks: { [key: string]: string }
+) {
+  // read file content
+  let fileContent = await fs.promises.readFile(filePath, "utf-8");
+  let fileChanged = false;
+
+  ({ fileContent, fileChanged } = changeFileContent(
+    fileContent,
+    changedPermalinks
+  ));
+
+  if (fileChanged) {
+    await fs.promises.writeFile(filePath, fileContent, "utf-8");
+    await simpleGit(repoPath).add(filePath);
+  }
+}
+
+function changeFileContent(
+  fileContent: string,
+  changedPermalinks: { [key: string]: string }
+) {
+  let fileChanged = false;
+  // three different regex patterns to take care of
+  // 1. href="original_permalink"
+  // 2. [click here](original_permalink)
```

Comment:
aye forgot to change this comment. Updated here: ef53d8c

Good question! Suppose we have a path that was changed from /ABOUT to /about. Say the page also has a link to an external website `www.googlesites.com/ABOUT`. If we did a naive `content.replace(`/ABOUT`,`about`)`, we would distruptively change their external link `www.googlesites.com/ABOUT` to `www.googlesites.com/about`. 

Therefore, there is a need to:
1) isolate the specific portions of the code where we know it is a permalink that is a of a relative path 
2) only change the links identified above

The general idea here is to _reduce_ the chances of changing something that we dont intend to, and only change the ones that we have a high confidence in changing!

Reply:
actually given what you've said earlier, that the url can only be specified in 2 forms, why can't we replace
`href="old_permalink"` and `(desc)[old_permalink]` instead, which avoids the clash

Reply:
Exactly what we are doing! But we cant just do a brute 
replace(`href="old_permalink"`, `href="new_permalink"`) since this would not account for cases like (list not exhaustive):
`href='old_permalink'` 
`href          ='old_peralink'`
`href          ="old_peralink"`
`href          =         "old_permalink"`

Thus, we would need to use `JSDOM` to generate the AST to then only modify the nodes that specifically refer to the `href` in the DOM!

Reply:
i think `href( )*=( )*("|')(.)+("|')` as a regex captures what we're trying to achieve. thereafter, we'd only need to omit spaces then we can do a replace

<img width="368" alt="Screenshot 2023-03-13 at 2 12 48 PM" src="https://user-images.githubusercontent.com/44049504/224621564-43cbc894-7e89-415d-a777-73c51c272770.png">

Reply:
Sure, we could use regex expressions to partially parse occurrences of `href=old_permalink`. 
I went with this option of using JSDOM because: 
1. True parsing of HTML is not possible with regular expressions as regular expressions can only match regular languages but HTML is a context-free language and not a regular language. 
2. While JSDOM might be too much of a solution, implementing it was fairly easy (literally like 10 ish lines of code). 
3. Didn't want to reinvent the wheel. If a powerful parser already exists that allows me to parse anchor tags, why not use it 
 
As previously mentioned, I would have done a similar approach with parsing of markdown, but there certain quirks that did not allow for a smooth output of the code after it was parsed, so I ended up using the regex method to parse markdown hyperlinks!

Reply:
hmm, my concern here is just that `jsdom` doesn't parse our html properly cos it's a mix of markdown + html, which is why i think that the regex replacement method is safer (+ we can capture the form of the replacement precisely in regex). 

if you feel like the parsing isn't an issue (ie, that `jsdom` parses html + markdown properly), feel free to resolve this

Reply:
fair point, + its hard to find documentation on this as we dont need JSDOM to really parse markdown, as long as it doesn't break in parsing HTML code, its fine for our use case. As a sanity check I will will manually looking through the output of the migration for the larger websites. There are some side effects for this parsing that I have noticed as well, eg, it will add closing tags for unclosed tags, effectively becoming a loose synthetic checker. Will change to use a regex pattern if got unintended behaviour!